dataclients/kubernetes/kube.go

```diff
@@ -79,37 +92,130 @@ type Options struct {
-	errServicePortNotFound = errors.New("service port not found")
+	errServiceNotFound      = errors.New("service not found")
+	errServicePortNotFound  = errors.New("service port not found")
+	errAPIServerURLNotFound = errors.New("kubernetes API server URL could not be constructed from env vars")
+	errInvalidCertificate   = errors.New("discovered certificate is invalid")
 )
 
 // New creates and initializes a Kubernetes DataClient.
-func New(o Options) *Client {
-	if o.KubernetesURL == "" {
-		o.KubernetesURL = defaultKubernetesURL
+func New(o Options) (*Client, error) {
+	httpClient, err := buildHTTPClient(o.KubernetesInCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	apiURL, err := buildAPIURL(o)
+	if err != nil {
+		return nil, err
 	}
 
-	log.Debugf("kube client initialized with api address: %s; with healthcheck: %t",
-		o.KubernetesURL, o.ProvideHealthcheck)
+	token, err := readServiceAccountToken(o.KubernetesInCluster)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("running in-cluster: %t. api server url: %s. provide health check: %t", o.KubernetesInCluster, apiURL, o.ProvideHealthcheck)
+
 	return &Client{
-		apiURL:             o.KubernetesURL,
+		httpClient:         httpClient,
+		apiURL:             apiURL,
 		provideHealthcheck: o.ProvideHealthcheck,
 		current:            make(map[string]*eskip.Route),
+		token:              token,
+	}, nil
+}
+
+func readServiceAccountToken(inCluster bool) (string, error) {
+	if !inCluster {
+		return "", nil
+	}
+	bToken, err := ioutil.ReadFile(serviceAccountDir + serviceAccountTokenKey)
+	if err != nil {
+		return "", err
 	}
+	log.Debugf("service account token is found")
+
+	return string(bToken), nil
+}
+
+func buildHTTPClient(inCluster bool) (*http.Client, error) {
+	if !inCluster {
+		return http.DefaultClient, nil
+	}
+
+	rootCA, err := ioutil.ReadFile(serviceAccountDir + serviceAccountRootCAKey)
+	if err != nil {
+		return nil, err
+	}
+	certPool := x509.NewCertPool()
+	if !certPool.AppendCertsFromPEM(rootCA) {
+		return nil, errInvalidCertificate
+	}
+
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		RootCAs:    certPool,
+	}
+
+	transport := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+
+	return &http.Client{
+		Transport: transport,
+	}, nil
+}
+
+func buildAPIURL(o Options) (string, error) {
+	if !o.KubernetesInCluster {
+		if o.KubernetesURL == "" {
+			return defaultKubernetesURL, nil
+		}
+		return o.KubernetesURL, nil
+	}
+
+	host, port := os.Getenv(serviceHostEnvVar), os.Getenv(servicePortEnvVar)
+	if len(host) == 0 || len(port) == 0 {
+		return "", errAPIServerURLNotFound
+	}
+
+	return "https://" + net.JoinHostPort(host, port), nil
+}
+
+func (c *Client) createRequest(method, url string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.token != "" {
+		req.Header.Set("Authorization", "Bearer "+c.token)
```

Comment:
looks, ok. You can run `make fmt` to apply the same formatting as the ci expects, or run `make check-fmt` to just check if it is the same as what the ci expects. (The task runs `gofmt -w`, which has probably the same settings as most editor defaults, but worth checking before push.)